includes/classes/SupportMonitor/DBQueryMonitor.php

```diff
@@ -0,0 +1,173 @@
+<?php
+/**
+ * Database Queries Monitor. A submodule of Support Monitor to report heavy SQL queries executed on staging.
+ *
+ * @since  x.x
+ * @package 10up-experience
+ */
+
+namespace TenUpExperience\SupportMonitor;
+
+/**
+ * DBQueryMonitor class
+ */
+class DBQueryMonitor {
+
+	/**
+	 * The transient name
+	 *
+	 * This transient stores all the queries logged.
+	 */
+	const TRANSIENT_NAME = 'tenup_experience_db_queries';
+
+	/**
+	 * Setup module
+	 */
+	public function setup() {
+		$production_environment = Monitor::instance()->get_setting( 'production_environment' );
+		if ( 'no' === $production_environment || apply_filters( 'tenup_experience_log_heavy_queries', false ) ) {
+			add_filter( 'query', [ $this, 'maybe_log_query' ] );
+		}
+	}
+
+	/**
+	 * Conditionally log a query
+	 *
+	 * Get all potential heavy queries (CREATE, ALTER, etc.) and store it,
+	 * ignoring transients and options by default.
+	 *
+	 * @param string $query The SQL query.
+	 * @return string
+	 */
+	public function maybe_log_query( $query ) {
+		global $wpdb;
+
+		if ( ! preg_match( '/^\s*(create|alter|truncate|drop|insert|delete|update|replace)\s/i', $query ) ) {
+			return $query;
+		}
+
+		if ( false !== strpos( $query, 'transient_' ) ) {
+			return $query;
+		}
+
+		if ( false !== strpos( $query, $wpdb->options ) ) {
+			return $query;
+		}
+
+		if ( apply_filters( 'tenup_experience_log_query', true, $query ) ) {
```

Comment:
@felipeelia Minor, but we should check for this as the first line in this function (checking for false) since if we don't want to log queries, we might as well not bother with the other checks and can bail early on.

Reply:
@bengreeley if people don't want to log queries at all, they should use the `tenup_experience_log_heavy_queries` filter, so the plugin doesn't even enter in the method. The intent of this `tenup_experience_log_query` is just to give the dev a way to refine things even further ("I don't want to log any queries related to the X plugin", for example). That said, if we move this filter to the beginning of the method, you'd prefer having it logging queries if the filter value is true *and* if it passes the other checks or *or*?

Reply:
I see. In that case, some additional docblocks about what the filters are intended to be used for would be helpful. My assumption was that this would be to toggle whether or not the logging functionality was activated at all.